lib/posix-user: New functions and cleanup

marcrittinghaus opened the pull request:
<!--

Thank you for opening a new PR to the Unikraft Open Source Project!  We welcome
new changes, features, fixes, and more!  Please fill in this form to indicate
the status of your PR.  Please ensure you have read the contribution guidelines
before opening a new PR as this will cover the PR process:

  https://unikraft.org/docs/contributing/

-->

### Prerequisite checklist

<!--
Please mark items appropriately:
-->

 - [x] Read the [contribution guidelines](https://unikraft.org/docs/contributing/) regarding submitting new changes to the project;
 - [x] Tested your changes against relevant architectures and platforms;
 - [x] Ran the [`checkpatch.pl`](https://github.com/unikraft/unikraft/blob/staging/support/scripts/checkpatch.pl) on your commit series before opening this PR;
 - [ ] Updated relevant documentation.


### Base target

 - Architecture(s): [N/A]
 - Platform(s): [N/A]
 - Application(s): [N/A]


### Additional configuration

<!--
Please specify any additional configuration which is needed for this feature to
work or any new configuration parameters which are introduced by this PR.  This
will help during the review process.  For example:

 - `CONFIG_LIBUKDEBUG=y`

-->

### Description of changes

The current `posix-user` implementation uses a lot of stubs and needs some cleanup as noted by the original authors:
![image](https://user-images.githubusercontent.com/8928900/179966575-ed674634-32d9-4403-887d-2e637d882982.png)

This PR performs the cleanup by
1. separating user/passwd and group related functions
2. ensuring consistent implementation schemes (e.g., using static global variable instead of iterator)
3. fixing minor formatting issues
4. harmonizing the order in which functions are implemented and referenced

From a functional perspective the PR does the following changes:
1. adds the following functions or their implementation: `login_r`, `getpwnam_r`, `getpwuid_r`, `getgrgid_r`, `getgrouplist`
2. fixes a bug in `getgrnam_r` where not enough memory is allocated to accommodate the array of group name pointers.
3. returns proper default values (e.g., `UK_DEFAULT_GID=0`) instead of just `0`. Although practically this does not make a difference
4. adds checks to all functions to accept only valid user/group IDs as it was already done for `setresgid`

Additional note: Whereas all functions make directly use of the fact that there is only a single passwd/group entry with fixed values, we do not assume the same in the `getgr_r` and `getpw_r` functions. This is due to the fact that these two functions are more complex to implement and we want to make sure these do not need to be changed IFF we would have multiple entries at some point (e.g., impersonation).

### Future Work

There are the `capget` and `capset` system calls which are currently implemented in `posix-user` although they are probably better moved to `posix-process` as they are related to the capabilities of threads. For now, we keep them here so that we do not collide with changes to `posix-process` that are about to be published by @skuenzer for musl compatibility.

### Dependencies

Before this PR can be merged, PR https://github.com/unikraft/lib-newlib/pull/22 needs to be merged ⚠️

marcrittinghaus committed:
lib/posix-user: Separate user and group calls

For a better overview this commit separates functions related to
user/password and group management and moves them into
their own files. The former `user.c` is used as `posix-user` main
that holds the constructor.

Signed-off-by: Marc Rittinghaus <marc.rittinghaus@kit.edu>

marcrittinghaus committed:
lib/posix-user: Fix signature formating

This commit fixes some of the function signatures and adds
comments for some of the functions to signal that not defining
this function a system call is by intention (i.e., it is not a syscall).

Signed-off-by: Marc Rittinghaus <marc.rittinghaus@kit.edu>

marcrittinghaus committed:
lib/posix-user: Harmonize group and user code

The user/passwd management functions currently use the
iterator interface also internally. However, as we only have
a single user, this is a waste of cycles. We thus do it like
for groups and make the default passwd entry static global
so we can work directly with it. The commit also changes
variable names to make passwd and group code more
similar.

Signed-off-by: Marc Rittinghaus <marc.rittinghaus@kit.edu>

marcrittinghaus committed:
lib/posix-user: Return proper default values

The current code widely just returns `0`. With this commit, we
return proper default values using the respective definitions.

Signed-off-by: Marc Rittinghaus <marc.rittinghaus@kit.edu>

marcrittinghaus committed:
lib/posix-user: Accept only valid values

The current code for setter mostly just return `0` no matter if the
supplied user or group exists. However, this is not the case for
`setresgid` which accepts only the default group. We adopt this
also for all setters.

Signed-off-by: Marc Rittinghaus <marc.rittinghaus@kit.edu>

marcrittinghaus committed:
lib/posix-user: Add getlogin_r function

This commit adds an implementation of the getlogin_r
function.

Signed-off-by: Marc Rittinghaus <marc.rittinghaus@kit.edu>

marcrittinghaus committed:
lib/posix-user: Add getpwnam_r/getpwuid_r

This commit removes the stub for getpwnam_r and getpwuid_r with
an actual implementation.

Signed-off-by: Marc Rittinghaus <marc.rittinghaus@kit.edu>

marcrittinghaus committed:
lib/posix-user: Fix getgrnam_r and add getgrgid_r

The current implementation of getgrnam_r has a bug as it does not
reserve memory for the actual array but only for the group names.
This commit fixes the bug and generalizes the implementation by
moving it do a dedicated function. This way, we can easily add
getgrgid_r.

Signed-off-by: Marc Rittinghaus <marc.rittinghaus@kit.edu>

marcrittinghaus committed:
lib/posix-user: Add getgrouplist

This commit adds an implementation for the getgrouplist function.

Signed-off-by: Marc Rittinghaus <marc.rittinghaus@kit.edu>

marcrittinghaus committed:
lib/posix-user: Cleanup Makefile and exportsyms

This commit cleans up the Makefile and exportsyms to reflect the order
in the implementation and add all new functions.

Signed-off-by: Marc Rittinghaus <marc.rittinghaus@kit.edu>

marcrittinghaus committed:
lib/posix-user: Temporarily reintroduce capget/set

The capset and capget system calls should probably be moved somewhere
else. As these calls were previously available here, we add them for
compatibility at this point.

Signed-off-by: Marc Rittinghaus <marc.rittinghaus@kit.edu>

StefanJum commented on lib/posix-user/group.c:
```diff
@@ -0,0 +1,309 @@
+	groups[0] = pwd->pw_gid;
+	if (group != pwd->pw_gid)
+		groups[1] = group;
+
```
```suggestion
*ngroups = ngrps;
```
Should we update the value stored in `ngroups` to match the actual number of groups found, even if it is greater or equal to `ngrps`?
From the [getgrouplist manual](https://linux.die.net/man/3/getgrouplist):
```text
The ngroups argument is a value-result argument: on return it always contains the number of groups
found for user, including group; this value may be greater than the number of groups stored in groups.
```

marcrittinghaus replied:
Yes, makes sense. The part of the man page describing the result value is a bit unclear but I assume you are right. I updated the code. Very good catch 😃

StefanJum commented on lib/posix-user/group.c:
```diff
@@ -0,0 +1,309 @@
+
+UK_SYSCALL_R_DEFINE(int, getgroups, int, size, gid_t *, list)
+{
+	return 0;
```
```suggestion
	if (size <= 0)
	    return 0;
	
	UK_ASSERT(list);
	list[0] = UK_DEFAULT_GID;
	
	return 1;
```

Should this return a list containing `UK_DEFAULT_GID`?

marcrittinghaus replied:
Actually, `initgroups`, `getgroups`, and `setgroups` are not implemented but only stubbed. But anyways, these work on supplementary groups so this would not include the default GID, but only additional ones. Since the specification does not state if the effective group ID is returned, it is fine to just return an empty list. However, I added a commit that "implements" `setgroups` so it returns `EINVAL` if an invalid GID is attempted to be set by the user. We then assume that even if the user specified a list containing only the default (valid) GID, we deduplicate this list against the effective GID. This way the list completely collapses and it is fine to return an empty list in `getgroups`.

StefanJum replied:
Ok, that makes a lot of sense, thanks.

StefanJum reviewed (commented):
Looks good besides the few comments.

marcrittinghaus committed:
lib/posix-user: Add implementation of s/getgroups

This commit changes the stubbed getgroups/setgroups to do more
meaningful operations. Since we only have a single group and this group
does not need to be returned in the list, we can effectively
ignore supplementary groups. However, we ensure that when setting
supplementary groups, only the default GID (i.e., a valid GID) is
attempted to be set so it is fine for getgroups to just return 0.

Signed-off-by: Marc Rittinghaus <marc.rittinghaus@kit.edu>

StefanJum reviewed (approved):
All looks good now.

StefanJum reviewed (changes requested):
Sorry I missed that earlier, but because of the moving of `user.h`, the build on some apps fail.
For example, when building `nginx`, we get the error `uk/user.h: No such file or directory`.

I think the `user.h` file should be moved into `posix-user/include/uk` and then the guards from `group.c` and `passwd.c` changed from `#include "user.h"` into `#include <uk/user.h>`.

marcrittinghaus commented:
Hi @StefanJum, thanks for pointing that out! 👍🏻 I must have missed this or fixed it for me and forgot about it. Could you state where the offending include is done?

I am against keeping `uk/user.h`, however, because it is introducing the signatures of glibc APIs and that has nothing to do in a Unikraft header. For `libnewlib` the declarations are missing, for `musl` they are in `unistd.h` with `_GNU_SOURCE` defined (as it should be).

StefanJum commented:
> Hi @StefanJum, thanks for pointed that out! 👍🏻 I must have missed this or fixed it for me and forgot about it. Could you state where the offending include is done?
> 
> I am against keeping `uk/user.h`, however, because it is introducing the signatures of glibc APIs and that has nothing to do in a Unikraft header. For `libnewlib` the declarations are missing, for `musl` they are in `unistd.h` with `_GNU_SOURCE` defined (as it should be).

Hi @marcrittinghaus, [this is the line](https://github.com/unikraft/lib-newlib/blob/staging/include/sys/unistd.h#L40) that causes the crash. It is the only one I've found.
It makes a lot of sense to not keep `uk/user.h`, you're right.

marcrittinghaus commented:
> Hi @marcrittinghaus, [this is the line](https://github.com/unikraft/lib-newlib/blob/staging/include/sys/unistd.h#L40) that causes the crash. It is the only one I've found. It makes a lot of sense to not keep `uk/user.h`, you're right.

Great! Thanks for looking this up. Actually, I can remember that I removed it locally. As the upcoming release will not be 100% musl yet, I will add a PR for `newlib` that fixes this. I have to create a `newlib` PR also for a different compatibility issue with the new eventpoll stuff anyways. I will reference the PR here.

StefanJum commented:
> 
> Great! Thanks for looking this up. Actually, I can remember that I removed it locally. As the upcoming release will not be 100% musl yet, I will add a PR for `newlib` that fixes this. I have to create a `newlib` PR also for a different compatibility issue with the new eventpoll stuff anyways. I will reference the PR here.

That's great, you can add me as a reviewer there then we can merge both PRs at the same time.

marcrittinghaus commented:
> That's great, you can add me as a reviewer there then we can merge both PRs at the same time.

@StefanJum I created the PR for the `uk/user.h` header. See the PR description for the reference. However, I am not able to assign you as reviewer in `lib-newlib`. Could you just test if it now works on your side, anyways?

StefanJum reviewed (approved):
@marcrittinghaus done. It all works fine now.

marcrittinghaus commented:
@razvand Should someone else be assigned to this so this makes it into the release?

razvand commented:
> @razvand Should someone else be assigned to this so this makes it into the release?

@marcrittinghaus , I'm assigning myself. Will check it out.

razvand commented on lib/posix-user/user.h:
```diff
@@ -1,8 +1,12 @@
 /* SPDX-License-Identifier: BSD-3-Clause */
 /*
- * Authors: Costin Lupu <costin.lupu@cs.pub.ro>
```
Any reason for Costin Lupu to be removed as author? His contributions have been removed from the code?

marcrittinghaus replied:
This is actually a new file and I just copied over the header. I removed Felipe and Simon as authors.

razvand replied:
roger

razvand commented on lib/posix-user/passwd.c:
```diff
@@ -0,0 +1,290 @@
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "user.h"
```
Why is this `#include "user.h"` and not `#include <uk/user.h>`?

marcrittinghaus replied:
See other comment.

razvand commented on lib/posix-user/group.c:
```diff
@@ -0,0 +1,333 @@
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "user.h"
```
Why is this `#include "user.h"` and not `#include <uk/user.h>`?

marcrittinghaus replied:
Because the header is no longer in the `include/uk` folder. Actually the old header no longer exists as well as the directory, as you pointed out in a different comment. 

This new header is private and I use `"` so the preprocessor looks in the same directory as the source file and not in the include directories.

razvand commented on lib/posix-user/exportsyms.uk:
```diff
@@ -1,101 +1,76 @@
-getpid
```
I'm not sure how these came to be here (`getpid` and `getppid`) but they are also not defined in any other place.

marcrittinghaus replied:
Yes, these do not belong here. They are defined in `posix-process` and exported there as no `exportsyms.uk` means all global symbols are exported.

razvand reviewed (changes requested):
Hi, @marcrittinghaus . Looks good, thanks for this work.

See my comments inline.

Other than those, there is an inconsistency in [this commit](https://github.com/unikraft/unikraft/pull/504/commits/1ef45afd873fc3fc24379de7b2b50e5166fe5497). The commit message says the function name is `login_r` but the actual function defined is `getlogin_r`.

Also, there is an issue with newlib, caused by removing the `include/user.h` file and moving it under `user.h` (local use only in the library):

```
In file included from /home/razvan/projects/unicore/unikraft-my/apps/app-redis.git/build/libnewlibc/origin/newlib-2.5.0.20170922/newlib/libc/include/unistd.h:4,
                 from /home/razvan/projects/unicore/unikraft-my/unikraft.git/plat/drivers/virtio/virtio_9p.c:34:
/home/razvan/projects/unicore/unikraft-my/libs/newlib/include/sys/unistd.h:40:10: fatal error: uk/user.h: No such file or directory
 #include <uk/user.h>
```

Also, the build command still includes the `lib/posix-user/include` folder. If it's no longer used it should be removed.

marcrittinghaus commented:
@razvand 
> See my comments inline.

Fixed and answered 👍🏻

> Other than those, there is an inconsistency in [this commit](https://github.com/unikraft/unikraft/pull/504/commits/1ef45afd873fc3fc24379de7b2b50e5166fe5497). The commit message says the function name is `login_r` but the actual function defined is `getlogin_r`.

Fixed 👍🏻

> Also, there is an issue with newlib, caused by removing the `include/user.h` file and moving it under `user.h` (local use only in the library):
> 
> ```
> In file included from /home/razvan/projects/unicore/unikraft-my/apps/app-redis.git/build/libnewlibc/origin/newlib-2.5.0.20170922/newlib/libc/include/unistd.h:4,
>                  from /home/razvan/projects/unicore/unikraft-my/unikraft.git/plat/drivers/virtio/virtio_9p.c:34:
> /home/razvan/projects/unicore/unikraft-my/libs/newlib/include/sys/unistd.h:40:10: fatal error: uk/user.h: No such file or directory
>  #include <uk/user.h>
> ```

This issue has already been mentioned by @StefanJum and has been fixed by making this PR dependent on:

1. https://github.com/unikraft/lib-newlib/pull/22

> Also, the build command still includes the `lib/posix-user/include` folder. If it's no longer used it should be removed.

Removed 👍🏻

razvand commented:
> This issue has already been mentioned by @StefanJum and has been fixed by making this PR dependent on:
> https://github.com/unikraft/lib-newlib/pull/22

Gotcha. I reviewed that, all is OK.

@StefanJum , please mark the `...-by` message for this PR and then I can finalize it to upstream it.

StefanJum reviewed (approved):
All good.
Reviewed-by: Stefan Jumarea <stefanjumarea02@gmail.com>

razvand reviewed (approved):
Works. Thanks, @marcrittinghaus 

Reviewed-by: Razvan Deaconescu <razvan.deaconescu@cs.pub.ro>
Approved-by: Razvan Deaconescu <razvan.deaconescu@cs.pub.ro>
